### ISSUE TITLE
[featureRequest]

### ISSUE THREAD

**Issue by shouldsee** [pandas-dev/pandas#25973]
### Code

Can we have a method that returns an iterator of dictionary instead of tuples? Something like

```python
def iterdicts(self, **kwargs):
    it = (x.__dict__ for x in self.itertuples())
    return it
```

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.12.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-143-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: None.None

pandas: 0.24.2
pytest: 3.2.3
pip: 19.0.3
setuptools: 40.8.0
Cython: 0.25.2
numpy: 1.14.0
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 5.7.0
sphinx: 1.7.4
patsy: 0.5.1
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.2.3
openpyxl: None
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.5
lxml.etree: 3.5.0
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

**Comment by simonjayhawkins** [comment]
DataFrame.to_dict(orient='records') currently returns a list. An alternative maybe to return an iterator. Or, to avoid a breaking api change, an additional orient method that maybe also includes the index.

**Comment by mroeschke** [comment]
In addition to @simonjayhawkins suggestion:

You can using the `_asdict()` method on the namedtuples returned by `itertuples`

Or just create a dict from the other iterator methods:

```
In [13]: df
Out[13]:
     0    1    2
0  1.0  0.0  0.0
1  0.0  1.0  0.0
2  0.0  0.0  1.0

In [14]: for i, row in df.iterrows():
    ...:     row = row.to_dict()
    ...:     row.update({'index': i})
    ...:     print(row)
    ...:
{0: 1.0, 1: 0.0, 2: 0.0, 'index': 0}
{0: 0.0, 1: 1.0, 2: 0.0, 'index': 1}
{0: 0.0, 1: 0.0, 2: 1.0, 'index': 2}
```

Personally -1 on adding another iterator method.

**Comment by simonjayhawkins** [comment]
@mroeschke : I think the issue is that this case is a corner case that sits between the current options, named tuples do not handle Python keywords, underscores etc, itterrows does not preserver dtypes and to_dict(orient=records) does the dataFrame at once and doesn't give the index values. also .to_records is not an generator.

**Comment by WillAyd** [comment]
Having a dict iterator is going to have corner cases as well, like the fact that we allow duplicate column names and don't require labels to be hashable

Iteration isn't something I think we really need to build a huge API out for with DataFrames, so I'd also be -1 here

**Comment by simonjayhawkins** [comment]
I agree that more methods is undesirable, but it may be worth keeping this issue open to discuss improvements to the current available methods to reduce the gaps.

for instance, the itertuples maybe should not rename by default and have a rename parameter, so that in the event of an invalid field name a friendly Exception is raised, if no exception safe to continue with creating dict, if exception call again with a rename parameter and proceed with caution.

see https://github.com/pandas-dev/pandas/issues/12799#issuecomment-479437753 for the driver behind this issue.

**Comment by shouldsee** [comment]
I did a bit of benchmarking. The result is in favour for an extension of iterrows()

```python
import pandas as pd
import string
import itertools

xrg = range(0,3)
xrga = np.array(list(string.ascii_lowercase))[xrg]

index = pd.MultiIndex.from_product([xrga,xrg])
columns = pd.MultiIndex.from_product([xrg,xrga])


dfc = pd.DataFrame( index=index,columns=columns)
it  = itertools.product(index,columns)
for i,(idx,col) in enumerate(it):
    val = map(str,idx + col)
    val = ''.join(val)
    dfc.values.flat[i] = val
    
import json
def ppJson(d):
    d = {unicode(k):v for k,v in d.items()}
    print json.dumps(d,indent=4)


def test__iterows(dfc):
    it  = dfc.iterrows()
    res = next(it)[1].to_dict()
    return res

def test__itertuples(dfc):
    it = dfc.itertuples()
    return next(it)._asdict()


def test__to_dict(dfc):
    it  = dfc.to_dict(orient='records')
    it  = iter(it)
    return next(it)


for methodName in ['test__iterows','test__itertuples','test__to_dict']:
    print ('[method]%s'%methodName)
    method = locals().get(methodName)
    %timeit method(dfc)
    ppJson(method(dfc))


```
